### PR TITLE
add initial stream resume implementation with tests

### PR DESCRIPTION
**Pull request by paweldomas**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,7 +18,14 @@
   },
   "homepage": "https://github.com/strophe/strophejs-plugin-stream-management#readme",
   "devDependencies": {
-    "rollup": "^0.45.2"
+    "http-server": "^0.10.0",
```
Comment (virtuacoplenny) [inline package.json:21]: Are all these new deps being used?

File: src/strophe.stream-management.js [hunk 2]
```diff
@@ -133,17 +149,58 @@ Strophe.addConnectionPlugin('streamManagement', {
...
 		this._c.xmlOutput = this.xmlOutput.bind(this);
+
+		this._originalConnect = this._c.connect;
+		this._c.connect = this._interceptConnectArgs.bind(this);
+
+		this._originalDoDisconnect = this._c._doDisconnect;
+		this._c._doDisconnect = this._interceptDoDisconnect.bind(this);
+
+		this._originalDisconnect = this._c.disconnect;
+		this._c.disconnect = this._interceptDisconnect.bind(this);
```
Comment (virtuacoplenny) [inline src/strophe.stream-management.js:166]: I'm thinking the connection could be a new class or passed into a decorator, so a lot of this logic can be moved out of stream-management and into a class dedicated to adding resume ability to connection. What do you think?

File: src/strophe.stream-management.js [hunk 3]
```diff
@@ -164,7 +224,29 @@ Strophe.addConnectionPlugin('streamManagement', {
 			}
 
 			this._requestHandler = this._c.addHandler(this._handleServerRequestHandler.bind(this), this._NS, 'r');
+			this._ackHandler = this._c.addHandler(this._handleServerAck.bind(this), this._NS, 'a');
 			this._incomingHandler = this._c.addHandler(this._incomingStanzaHandler.bind(this));
+
+			// FIXME handler instances stored, but never used
```
Comment (virtuacoplenny) [inline src/strophe.stream-management.js:230]: Does this result in leaked handlers or handlers that would get called when they shouldn't due to a successful reconnect?

File: tests/tests.js [hunk 4]
```diff
@@ -0,0 +1,519 @@
+/* global equal, notEqual, ok, module, test */
+
+define([
```
Comment (virtuacoplenny) [inline tests/tests.js:3]: Did you go with require and qunit because strophe uses them?

Comment (paweldomas) [reply tests/tests.js:3]: yes

File: package.json (hunk 1 above)
Comment (paweldomas) [reply package.json:21]: almost all, I was having issues in making chrome headless running the tests, I guess I will remove all not used ones now and  re-add them back

Comment (paweldomas) [reply package.json:21]: just checked it now and they are all used

**Comment by paweldomas** [comment]
Opened off the feature branch https://github.com/jitsi/strophejs-plugin-stream-management/pull/2

**Pull request closed without merging**